### ISSUE TITLE
RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by luweikxy** [Khan/KaTeX#1713]
When I build gitbook, it shows that Template render error: ...RangeError: Maximum call stack size exceeded
![](https://wx1.sinaimg.cn/mw690/006f01Rgly1fv9hn5x5tlj30gj0940ta.jpg)
So I tryed many time to locate the place where may cause this error. I found that if I delete some of  math formula, then it is OK. but if I delete some math formula in other place, it is also OK. That is to say, the reason is not because my math formula is wrong.
=========
I am sure it is due to Katex, because in my book.json, if I forbid katex, all is OK. It is not que to lunr
or other plugins.

"plugins": [
		"katex@^1.1.3",

**Comment by edemaine** [comment]
@luweikxy Could you give an example of a formula that, when you delete it, fixes the error?

**Comment by luweikxy** [comment]
> @luweikxy Could you give an example of a formula that, when you delete it, fixes the error?

Ok, this is the formula that when I delete it ,it is ok:
![](https://wx4.sinaimg.cn/mw690/006f01Rgly1fva1agx8hcj31080l840x.jpg)
this is the corresponding markdown source code:
![](https://wx2.sinaimg.cn/mw690/006f01Rgly1fva1aguru2j30mc0hdt9g.jpg)

the .md I modify is:
https://github.com/luweikxy/machine-learning-notes/blob/master/content/machine-learning/support-vector-machine/support-vector-machine.md

**Comment by ronkok** [comment]
@luweikxy According to [this page](https://github.com/GitbookIO/plugin-katex/tree/master/static), GitBook's KaTeX plugin uses KaTeX version 0.5.1. A lot has changed in the last three years. The current version is [v0.10.0-rc.1](https://github.com/khan/katex/releases).  Would it be possible to upgrade the KaTeX plugin?

**Comment by luweikxy** [comment]
> @luweikxy According to [this page](https://github.com/GitbookIO/plugin-katex/tree/master/static), GitBook's KaTeX plugin uses KaTeX version 0.5.1. A lot has changed in the last three years. The current version is [v0.10.0-rc.1](https://github.com/khan/katex/releases). Would it be possible to upgrade the KaTeX plugin?

My katex version is 1.1.3 based my book.json:
"plugins": [
"katex@^1.1.3",

where can I know my version is 0.5.1?

and I update my last reply just now, could you have a look? does katext have a num limites?

**Comment by luweikxy** [comment]
> @luweikxy The plug-in version is 1.1.3. But that plugin has dependencies, and the dependencies have me confused. According to the plugin [package JSON](https://github.com/GitbookIO/plugin-katex/blob/master/package.json), it is dependent on KaTeX 0.7.1, but according to the plug-in [static file](https://github.com/GitbookIO/plugin-katex/tree/master/static), it is using the CSS and fonts from KaTeX version 0.5.1. That is a little odd, but it should not cause a stack overflow.
> 
> KaTeX has no limit on the number of formulas. The plug-in is making repeated calls to the `katex.renderToString` function, but KaTeX runs each one of those calls sequentially. There is no reason it could not go on indefinitely.
> 
> KaTeX does create quite a lot of HTML for each formula, and your file is quite large. It is possible that the results have met a GitBook size limit.

Ok, I will delete 4 formulas so as to not exceed the whatever size...

How can I upgrade the katex by myself? change the [package.json](https://github.com/GitbookIO/plugin-katex/blob/master/package.json) 
from
"dependencies": {
        "katex": "0.7.1"
    },
to 
"dependencies": {
        "katex": "0.10.0"
    },
in my node_modules folder?

**Comment by ronkok** [comment]
> How can I upgrade the katex by myself? change the package.json[?]

Yes, I suppose one could fork GitBook and do that update. Or you could write a pull request to GitBook so that everyone gets the benefit of the update.  

After running `npm update`, check the files in the [static folder](https://github.com/GitbookIO/plugin-katex/tree/master/static). They should also be updated to current versions.

**Comment by luweikxy** [comment]
> > How can I upgrade the katex by myself? change the package.json[?]
> 
> Yes, I suppose one could fork GitBook and do that update. Or you could write a pull request to GitBook so that everyone gets the benefit of the update.
> 
> After running `npm update`, check the files in the [static folder](https://github.com/GitbookIO/plugin-katex/tree/master/static). They should also be updated to current versions.

I've give an issue to the GitBook.
https://github.com/GitbookIO/plugin-katex/issues/22

**Comment by gaoxiaosong** [comment]
Because gitbook use nunjucks v2.5.2 and its algorithm has the problem when process many blocks. Please see this issue:

[Including template many times in forloop causes a `RangeError: Maximum call stack size exceeded`](https://github.com/mozilla/nunjucks/issues/643)

Nunjucks has fixed this problem in v3.0 and later, but gitbook has not been updated.

**Comment by gaoxiaosong** [comment]
You can split the markdown file to some smaller one. Make sure there is only a few blocks in each file. @luweikxy
